src/Interactive/EditorFeatures/CSharp/Interactive/CSharpSendToInteractiveSubmissionProvider.cs

```diff
@@ -0,0 +1,124 @@
+
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editor.Interactive;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.Interactive
+{
+    [Export(typeof(ISendToInteractiveSubmissionProvider))]
+    internal sealed class CSharpSendToInteractiveSubmissionProvider
+        : SendToInteractiveSubmissionProvider
+    {
+        protected override bool CanParseSubmission(string code)
+        {
+            SourceText sourceCode = SourceText.From(code);
+            ParseOptions options = CSharpParseOptions.Default.WithKind(SourceCodeKind.Script);
+            SyntaxTree tree = SyntaxFactory.ParseSyntaxTree(sourceCode, options);
+            if (tree == null)
+            {
+                return false;
+            }
+
+            return tree.HasCompilationUnitRoot && !tree.GetDiagnostics().Any();
+        }
+
+        protected override IEnumerable<TextSpan> GetExecutableSyntaxTreeNodeSelection(TextSpan selectionSpan, SourceText source, SyntaxNode root, SemanticModel model)
+        {
+            SyntaxNode expandedNode = GetExecutableSyntaxTreeNode(selectionSpan, source, root, model);
+            return expandedNode != null
+                ? new TextSpan[] { expandedNode.Span }
+                : Array.Empty<TextSpan>();
+        }
+
+        private SyntaxNode GetExecutableSyntaxTreeNode(TextSpan selectionSpan, SourceText source, SyntaxNode root, SemanticModel model)
+        {
+            Tuple<SyntaxToken, SyntaxToken> tokens = GetSelectedTokens(selectionSpan, root);
+            var startToken = tokens.Item1;
+            var endToken = tokens.Item2;
+            if (startToken != endToken && startToken.Span.End > endToken.SpanStart)
+            {
+                return null;
+            }
+
+            // If a selection falls within a single executable statement then execute that statement.
+            var startNode = GetGlobalExecutableStatement(startToken);
+            var endNode = GetGlobalExecutableStatement(endToken);
+            if (startNode == null || endNode == null)
+            {
+                return null;
+            }
+
+            // If one of the nodes is an ancestor of another node return that node.
+            if (startNode.Span.Contains(endNode.Span))
+            {
+                return startNode;
+            }
+            else if (endNode.Span.Contains(startNode.Span))
+            {
+                return endNode;
+            }
+
+            // Selection spans multiple statements.
+            // In this case find common parent and find a span of statements within that parent.
+            var commonNode = root.FindNode(TextSpan.FromBounds(startNode.Span.Start, endNode.Span.End));
```

Comment:
I suspect it would be cheaper to find their least-common ancestor, but I haven't reviewed the implementation of `FindNode`.